### PR TITLE
GROOVY-10942 Add `extension` & `baseName` extension methods

### PR DESCRIPTION
**Pull request by Goooler**
Ref:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io.path/java.nio.file.-path/extension.html
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io.path/java.nio.file.-path/name-without-extension.html

### CODE REVIEW

**Comment by paulk-asert** [comment]
I like the idea behind these extensions. Can you create a corresponding Jira issue?

Also, some minor points:

- I think using `getExtension` rather than `extension` would be closer to existing methods. WDYT?
- Also, `extension` doesn't currently return empty string when there is no extension. This is what the javadocs state but not what the implementation does. The javadoc matches how Kotlin and commons-io FilenameUtils behave.
- An alternative to `nameWithoutExtension` is `baseName` (what commons-io uses) or just `basename` (again, I'd go with the getter variant). I guess there is some merit in following Kotlin but `basename` would certainly be shorter. WDYT?

Naming is always tricky, we can always ask on the mailing list if we want to confirm preferences.

**Comment by Goooler** [comment]
GROOVY-10942

1. Agreed.
2. I misjudged `Path.getFileName()` to return a non-null String, will fix it later.
3. Will add a `getBaseName()`.

File: subprojects/groovy-nio/src/main/java/org/apache/groovy/nio/extensions/NioExtensions.java [hunk 1]
```diff
@@ -101,6 +101,36 @@ public static long size(Path self) throws IOException {
...
+    /**
+     * Provide the extension of this path.
+     *
+     * @param self a path object
+     * @return the extension of this path (not including the dot), or an empty string if it doesn't have one.
+     * @since 5.0.0
+     */
+    public static String getExtension(Path self) {
+        final Path fileName = self.getFileName();
+        final String name = fileName == null ? "" : fileName.toString();
```
Comment (Goooler) [inline subprojects/groovy-nio/src/main/java/org/apache/groovy/nio/extensions/NioExtensions.java:113]: If we can resue this part, add a `Path.name()` like [`Path.name` in Kotlin](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io.path/java.nio.file.-path/name.html)?

**Comment by paulk-asert** [comment]
Can you elaborate what you mean by "too implicit"? (Just checking I understood) You mean "too obscure" whereas the longer one is "more self-describing"? I guess the shorter one relies on some familiarity with *nix commands which might not always be the case.

File: subprojects/groovy-nio/src/main/java/org/apache/groovy/nio/extensions/NioExtensions.java (hunk 1 above)
Comment (paulk-asert) [reply subprojects/groovy-nio/src/main/java/org/apache/groovy/nio/extensions/NioExtensions.java:113]: You want `Path.getName()`, with shorthand `Path.name` as a shorthand for `Path.getFileName()`/`Path.fileName`?
That could work but there is the existing `getName(int index)` method which might confuse some folks.

Comment (Goooler) [reply subprojects/groovy-nio/src/main/java/org/apache/groovy/nio/extensions/NioExtensions.java:113]: Yeah, that's what I'm afraid.

**Comment by Goooler** [comment]
Renamed to `getBaseName`.

**Pull request merged**